hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go

```diff
@@ -0,0 +1,565 @@
+				actualDestination,
+				elemType,
+				elemIdent.Name,
+				append(conversionContext, destinationType),
+				astbuilder.AppendList),
+		},
+	}
+	results = append(results, result)
+	if depth > 0 {
+		results = append(results, assignmentHandler(destination, actualDestination))
+	}
+
+	return results
+}
+
+// convertComplexMapProperty handles conversion for map properties with complex values.
+// This function panics if the map keys are not primitive types.
+// This function generates code that looks like this:
+// 	if <source> != nil {
+//		<destination> = make(map[<destinationType.KeyType()]<destinationType.ValueType()>)
+//		for key, value := range <source> {
+// 			<code for producing result from destinationType.ValueType()>
+//			<destination>[key] = <result>
+//		}
+//	}
+func (builder *convertFromArmBuilder) convertComplexMapProperty(
+	source ast.Expr,
+	destination ast.Expr,
+	destinationType *astmodel.MapType,
+	_ string,
+	conversionContext []astmodel.Type,
+	assignmentHandler func(lhs ast.Expr, rhs ast.Expr) ast.Stmt) []ast.Stmt {
+
+	if _, ok := destinationType.KeyType().(*astmodel.PrimitiveType); !ok {
+		panic(fmt.Sprintf("map had non-primitive key type: %v", destinationType.KeyType()))
+	}
+
+	depth := countArraysAndMapsInConversionContext(conversionContext)
+
+	keyIdent := ast.NewIdent("key")
+	valueIdent := ast.NewIdent("value")
+	elemIdent := ast.NewIdent("elem")
+
+	actualDestination := destination // TODO: improve name
+	makeMapToken := token.ASSIGN
+	if depth > 0 {
+		actualDestination = elemIdent
+		elemIdent = ast.NewIdent(fmt.Sprintf("elem%d", depth))
+		makeMapToken = token.DEFINE
+	}
+
+	handler := func(lhs ast.Expr, rhs ast.Expr) ast.Stmt {
+		return astbuilder.InsertMap(lhs, keyIdent, rhs)
+	}
+
+	keyTypeAst := destinationType.KeyType().AsType(builder.codeGenerationContext)
+	valueTypeAst := destinationType.ValueType().AsType(builder.codeGenerationContext)
+
+	makeMapStatement := astbuilder.SimpleAssignment(
+		actualDestination,
+		makeMapToken,
+		astbuilder.MakeMap(keyTypeAst, valueTypeAst))
+	rangeStatement := &ast.RangeStmt{
+		Key:   keyIdent,
+		Value: valueIdent,
+		X:     source,
+		Tok:   token.DEFINE,
+		Body: &ast.BlockStmt{
+			List: builder.fromArmComplexPropertyConversion(
+				valueIdent,
+				actualDestination,
+				destinationType.ValueType(),
+				elemIdent.Name,
+				append(conversionContext, destinationType),
+				handler),
+		},
+	}
+
+	result := &ast.IfStmt{
+		Cond: &ast.BinaryExpr{
+			X:  source,
+			Op: token.NEQ,
+			Y:  ast.NewIdent("nil"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				makeMapStatement,
+				rangeStatement,
+			},
+		},
+	}
+
+	if depth > 0 {
+		result.Body.List = append(result.Body.List, assignmentHandler(destination, actualDestination))
+	}
+
+	return []ast.Stmt{result}
+}
+
+// convertComplexTypeNameProperty handles conversion of complex TypeName properties.
+// This function generates code that looks like this:
+//	<nameHint> := <destinationType>{}
```

Comment:
This has an issue which is that it won't work if `<destination>` is `nil`, at least with the way that these functions are written right now.

[My plan](https://github.com/Azure/k8s-infra/pull/147/files#diff-ca7eb123d2f0acd0ebafb4b1ace7e9cdR480) was that at the top level we'd do this:
```
    objFromAzure := scheme.new(gvk) // We need to provide the empty type to deserialize into
    // Somehow construct a new object of type etcdObject
    if armTransformer, ok := objFromAzure.(ARMTransformer); ok {
        result := armTransformer.ToArm("") // This just converts from an empty kube shape to an empty arm shape
        armClient.GetIt(armId, result)
    
        armTransformer.FromArm(ownerRef, result)
    }
```

I think if we weren't concerned about reusing an object that already had stuff in it maybe we could do what you're suggesting, but I think we are concerned about that?

Reply:
Isn't that loading into an empty object though? And presumably if the `FromArm` call returns an error, we'll stop (at least this time around the loop), discarding the data from arm? (This is not a biggie either way, just trying to understand.)

I've tried to follow the linked reconciliation code and I think there's a step I'm missing - there doesn't seem to be a connection between the object retrieved from arm (`objFromAzure`) and the one we send back (`resource`). Is that just because we assume the update is idempotent, so we don't need to compare the desired state to the current state, we just always write the desired state?

Reply:
Let's sync up on this in person on Monday PST.

Reply:
As discussed in person, for non-optional types right right, we can do what you suggest.
For optional types, because the default is nil we need to somehow construct a container to hold the result, fill that result out, and then assign a ptr to it to the final variable. That ends up looking like this:
```
        err = optionalFoo.FromArm(owner, typedInput.OptionalFoo)
        if err != nil {
            return err
        }
        fakeResourceSpec.OptionalFoo = &optionalFoo
```
Because of the recursive nature of the functions, it's easier to just have the "standard case" and this "optional case" very close-together in terms of code structure, so that we can embed one in the other easily.

Reply:
Resolving this as @babbageclunk agreed that this makes sense.